# The Chi-square Tests

In this notebook we will learn about two types of Chi-square tests.  The Goodness-of-Fit test and the Test for Independence.  This may be an overgeneralization but in my opinion the biggest difference between the two is how many categorical variables are involved.  The goodness-of-fit test uses just one categorical variable, but the test for independence uses two categorical variables.  

In [1]:
# This cell is a code cell.  In cells like this
# we run Python code

# Here's some code that will likely appear near the top of every homework or lecture this semester.

from datascience import *
import numpy as np

%matplotlib inline
import matplotlib.pyplot as plots
plots.style.use('fivethirtyeight')

import scipy.stats as stats

import scipy

import warnings
warnings.simplefilter(action="ignore", category=FutureWarning)
warnings.simplefilter(action="ignore", category=np.VisibleDeprecationWarning)

# To add text, referred to as comments to a code cell
# just put a hashtag a the beginning of the comment
# Comments are ignored by the computer when executing the code

mort22 = Table.read_table("SmallMort2022.csv")

From a survey at Pew Research Center, we find the following:


|     |Right/Leans Right  |  Left/Leans Left  |
|:-:  |  :-:              |  :-:              |
|Men  |2343|2073|
|Women|2444|2833|


# Goodness-of-Fit

Is it the case that roughly 50% of the country leans each way, politically?

We can use this Pew Research Center data and the chi-square goodness-of-fit test to address that question.  

The null hypothesis is that, Yes, 50% will lean Right and 50% will lean left.

$H_o: \pi = \left(0.5, 0.5\right) $

$H_a: \pi \not= \left(0.5, 0.5\right) $

Unless we discuss it ahead of time and decide to change our $\alpha$, let's always assume that $\alpha = 0.05$.

The reason we may change $\alpha$ would be to limit our exposure to either Type I or Type II Error.  But that's not relevant now, let's move on.  


The command is `chisquare` and is included in the `scipy.stats` package.  In the cell below, we access it by it's full name, but as you know, we can also give it a new, shorter name as part of an import statement.  

The `chisquare` function needs us to give the observed data followed by what we'd expect the data to look like if the null hypothesis is true.  

For the observed, use the counts found in your real data.  For the expected, first, let $\pi$ be the vector of proportions/probabilities that are a mathematical representation of the null hypothesis.  In this case, $\pi = \left(\frac{1}{2}, \frac{1}{2} \right)$.  We always need that $\sum \pi = 1$, by the way.  Then, use the observed data again, but just to get the total sample size, $N$.  Then expected $ = N\pi$.


In [2]:
observed = np.array([2343+2444, 2073+2833])

expected = np.array([0.5, 0.5]) * (2343+2444+2073+2833)

scipy.stats.chisquare(observed, expected)

Power_divergenceResult(statistic=1.4609512018982771, pvalue=0.22677883522068867)

In [3]:
observed

array([4787, 4906])

In [4]:
expected

array([ 4846.5,  4846.5])

**Report** We see that the p-value is 0.2268, which is larger than our preselected $\alpha = 0.05$.  Therefore, at the 5% level of significance, we conclude that this data does NOT support the claim that either political leaning is more popular than the other.  Essentially, this data supports the claim that 50% leans right and 50% leans left.  

What is this test doing?

$\displaystyle \chi^2 = \sum_{i=1}^n \frac{\left(O_i-E_i\right)^2}{E_i}$

$\displaystyle \frac{(4787 - 4846.5)^2}{4846.5} + \frac{(4906 - 4846.5)^2}{4846.5} \approx 1.46095$

Then the p-value comes from the $\chi^2$ distribution with degrees of freedom equal to one less than the number of groups involved; so in this case, the degrees of freedom are just 1.  


In [5]:
(4787 - 4846.5)**2/4846.5 + (4906 - 4846.5)**2 / 4846.5

1.460951201898277

In [6]:
1 - scipy.stats.chi2.cdf(1.4609512, 1)

0.22677883552248579

Let's look at the Pew data split by gender.  Are men equally likely to lean left as to right?  Are women?

In [7]:
observed = np.array([2343, 2073])

expected = np.array([0.5, 0.5]) * (2343+2073)


scipy.stats.chisquare(observed, expected)

Power_divergenceResult(statistic=16.508152173913043, pvalue=4.8441349580536515e-05)

**Reporting out:** With a p-value of approximately $4.8\times 10^{-5}$ or $p = 0.000$, we conclude that men do not split evenly between leaning left and right.  Because there are only two categories in this data, it is a simple matter to look closer at the numbers and realize that men have a tendency to lean right.  Depending upon the audience, you may include the statistic and p-value in your concluding statement.  Men tend to lean politically right ($\chi^2 = 16.51$, $p = 0.000$). 

In [8]:
observed = np.array([2444, 2833])

expected = np.array([0.5, 0.5]) * (2444+2833)


scipy.stats.chisquare(observed, expected)

Power_divergenceResult(statistic=28.675573242372561, pvalue=8.5576378371485734e-08)

**Reporting out:** With a p-value of approximately $p = 0.000$, we conclude that women do not split evenly between leaning left and right.  Women tend to lean politically left ($\chi^2 = 28.68$, $p = 0.000$). 

### Another Example

Are people dying their hair different colors?  According to [World Population Review](https://worldpopulationreview.com/country-rankings/hair-color-by-country), 85% of Americans naturally have black hair, 11% have naturally brown hair, 2% have naturally blonde hair and 1% have naturally red hair.  Ahem, I'm sure you noticed that these values only sum to 99%; while we're not certain this could be due to rounding error or omitting certain very rare colors--or most likely, a combination of both factors.  

Let's divide that extra 1% between the four groups.

$\pi = \left(0.8525, 0.1125, 0.0225, 0.0125\right)$

Suppose we look around a crowded room of college students and (by our own judgement) count the following hair colors:

|Color|Number|
|---|:-:|
|Black |20 |
|Brown | 4|
|Blonde|10|
|Red   |6|

Is there evidence that, for one reason or another, this distribution of hair colors does not conform to the prediction from World Review?

We'll run the goodness-of-fit test below.  


In [37]:
observed = np.array([20, 4, 10, 6])

pi = np.array([0.8525, 0.1125, 0.0225, 0.0125])

expected = pi * sum(observed)


scipy.stats.chisquare(observed, expected)

Power_divergenceResult(statistic=158.39687194525902, pvalue=4.0652377672889614e-34)

In [38]:
100*observed/sum(observed)

array([ 50.,  10.,  25.,  15.])

**Reporting out:** Based on this data, this group of people does not fit the predicted distribution of hair colors, ($\chi^2 = 158.40$, $p = 0.000$). Looking at the actual percentage in the considered data we see that only 50% of the people have black hair which is less than predicted, while blonde and red hair are both greatly overrepresented at 25% and 15%, respectively.  

# Independence 

$\chi^2$ Test for independence

To test for a relationship of some pair of categorical variables, we can use this test for independence.

In [9]:
mort22.pivot("Suicide", "Sex")

Sex  | No     | Yes
F    | 620928 | 4138
M    | 674905 | 15856

In [10]:
obs = np.array([[620928, 4138],[674905, 15856]])

stats.chi2_contingency(obs)

(5849.5627113705341,
 0.0,
 1,
 array([[ 615568.11798056,    9497.88201944],
        [ 680264.88201944,   10496.11798056]]))

In [11]:
(19994 *625066 )/ 1315827**2

0.007218184472156906

The expected matrix (despite being the final part of the output) is actually the first thing created.  How?  First, the margins are summed.

|Sex|  No|Yes| |
|--:| --:|--:|--|
|F|620928 |4138|620928 + 4138|
|M|674905 |15856|674905 + 15856 |        
| |620928+674905 |4138+15856|       |



|Sex|  No|Yes| |
|--:| --:|--:|--|
|F|620928 |4138|625066|
|M|674905 |15856|690761 |        
| |1295833 |19994|    |

Then the table total is computed too.  

$N = 620928+674905 + 4138+15856 = 1315827$

Then these marginal totals and the table total are used to compute the probability of landing in each cell of the table, assuming that these are independent.  


|Sex|  No|Yes|
|--:| --:|--:|
|F|$$\frac{1295833}{1315827} \cdot \frac{625066}{1315827}$$ |$$\frac{19994}{1315827} \cdot \frac{625066}{1315827}$$|
|M|$$\frac{1295833}{1315827} \cdot \frac{690761}{1315827}$$ |$$\frac{19994}{1315827} \cdot \frac{690761}{1315827}$$ |        

|Sex|  No|Yes|
|--:| --:|--:|
|F|0.4678 |0.0072 |
|M|0.5170 |0.0079 |        


Then, these probabilities are converted back into counts by multiplying by N.

|Sex|  No|Yes|
|--:| --:|--:|
|F|615568.12 |9497.88 |
|M|680264.88 |10496.12| 


Then the observed values are $O_i$ and these expected values are denoted $E_i$.  


Then $\displaystyle \chi^2 = \sum_{i=1}^n \frac{\left(O_i-E_i\right)^2}{E_i}$

In this case, $n = 4$, but it's the product of the number of row and the number of columns in your two-way table. 

It's not necessary, but in the cell below we'll rerun the test so we can examine the output without having to scroll back up.  


In [12]:
stats.chi2_contingency(obs)

(5849.5627113705341,
 0.0,
 1,
 array([[ 615568.11798056,    9497.88201944],
        [ 680264.88201944,   10496.11798056]]))

How can we access the individual parts of this output?

There are (at least) two ways.  That is, there are two ways that I typically use.

**Method 1:** Name an array and set it equal to the test results, then pull the individual parts by position number.  


**Method 2:** Create a tuple with the names you want to use and set that tuple equal to the test, then call the parts by the name you gave.  

In [13]:
results = stats.chi2_contingency(obs)

In [14]:
results[0]

5849.5627113705341

In [15]:
results[1]

0.0

In [16]:
results[2]

1

In [17]:
results[3]

array([[ 615568.11798056,    9497.88201944],
       [ 680264.88201944,   10496.11798056]])

In [18]:
(X2, p, df, exp_tab) = stats.chi2_contingency(obs)

In [19]:
X2

5849.5627113705341

In [20]:
p

0.0

In [21]:
df

1

In [22]:
exp_tab

array([[ 615568.11798056,    9497.88201944],
       [ 680264.88201944,   10496.11798056]])

In [23]:
bf = Table.read_table("bf_summary_table.csv")

bf

season  | Midwest | Northeast | Southeast | West
Fall    | 362     | 70        | 376       | 399
Spring  | 197     | 39        | 212       | 220
Summer  | 407     | 76        | 418       | 586
Unknown | 17      | 1         | 29        | 17
Winter  | 164     | 35        | 249       | 171

In [24]:
bf_table = np.array([[362, 70, 376, 399],
                    [197,  39, 212, 220],
                    [407,  76, 418, 586],
                    [164,  35, 249, 171]])


stats.chi2_contingency(bf_table)

(43.920754335935875,
 1.4596890966788743e-06,
 9,
 array([[ 342.60487315,   66.70183371,  380.5036423 ,  417.18965084],
        [ 189.61065059,   36.9153479 ,  210.58528008,  230.88872143],
        [ 422.08239136,   82.17533283,  468.77292138,  513.96935443],
        [ 175.7020849 ,   34.20748556,  195.13815624,  213.9522733 ]]))

In [25]:
(x2, p, df, exp_tab) = stats.chi2_contingency(bf_table)

## The residuals

Just like with linear regression, the difference between observed and expected/predicted is the residual. So the $(i, j)^{th}$ residual is $o_{ij} - e_{ij}$.  



In [26]:
residuals = bf_table - exp_tab

residuals

array([[ 19.39512685,   3.29816629,  -4.5036423 , -18.18965084],
       [  7.38934941,   2.0846521 ,   1.41471992, -10.88872143],
       [-15.08239136,  -6.17533283, -50.77292138,  72.03064557],
       [-11.7020849 ,   0.79251444,  53.86184376, -42.9522733 ]])

If we divide the residuals by the standard deviation of the observations, we get what is known as the Pearson residuals.  

If we divide the residuals by their own standard deviations, we get the standardized residuals.  What is their own standard deviation?

$$\sqrt{e_{ij}(1-p_i)(1-p_j)}$$ where $p_i$ and $p_j$ are the marginal probabilities.  

Unfortunately, `scipy.stats` does not currently have a built-in function for computing these more advanced residuals, but `statsmodels` does, so we'll use theirs.  

In [53]:
import statsmodels.api as sm                                                                         

table = sm.stats.Table(bf_table)                                                                            

table.resid_pearson  # Pearson's residuals

array([[ 1.48332459,  0.49772671, -0.3342416 , -1.31884174],
       [ 0.69511665,  0.3869545 ,  0.12914423, -0.97107578],
       [-1.0960161 , -0.88548645, -3.58040308,  4.9623659 ],
       [-1.13519247,  0.15170094,  5.07038007, -3.9501825 ]])

In [54]:
table.standardized_resids  # Standardized residuals

array([[ 1.48332459,  0.49772671, -0.3342416 , -1.31884174],
       [ 0.69511665,  0.3869545 ,  0.12914423, -0.97107578],
       [-1.0960161 , -0.88548645, -3.58040308,  4.9623659 ],
       [-1.13519247,  0.15170094,  5.07038007, -3.9501825 ]])

Those look the same! And, and which one should I use in practice.  First, of all they do look the same, which is actually okay because they are trying to serve the same purpose.  If they were radically different, that would put us in an awkward situation.

Because the code is somewhat "blackbox", I don't always know what math the computer is doing.  In particular, I don't know what how the computer is finding the standard error for the Pearson residuals, I could try find out, but I'm not going to do that now.  I do know how the standardized residuals are found, so we can talk about that.    


Since they are standardized, that is divided by something similar to their own standard deviations, that means when one of them is smaller than -3 or larger than +3, that indicates that it is unusually small (or large).  

Which one should I use?  Since they are probably roughly the same, from one point of view it doesn't matter which one you use.  However, I personally prefer to be able to explain my work, if I'm ever challenged by a peer or other stakeholder, so I would tend to use the standardized residual for the simple reason that if I had to, I could walk someone through the math of how they are calculated.

Having stated that, let's try to recreate just some of those calculations in the cells below.  First, do the easiest one to find, the top left corner.  Later, we can try the residual of 53.86, which has the highest standardized residual despite not being the largest residual.  

**By the way!!** I will not be asking you to do these by-hand calculations.  You just need to know how to look at a set of standardized residuals and pick out which ones are unusually large or small.  (Small sometimes means near 0, but in this case by small we mean large-but-negative.)


In [56]:
# First, we'll calculate the margin totals, there's a command for that

margins = scipy.stats.contingency.margins(bf_table)

margins

[array([[1207],
        [ 668],
        [1487],
        [ 619]]),
 array([[1130,  220, 1255, 1376]])]

In [64]:
margins[0]

array([[1207],
       [ 668],
       [1487],
       [ 619]])

In [81]:
#Get the (1-p_i)

Q_0 = 1 - margins[0]/sum(sum(margins[0]))


Q_0

array([[ 0.69680985],
       [ 0.83220296],
       [ 0.62647576],
       [ 0.84451143]])

In [82]:
# Get the (1-p_j)

Q_1 = 1-margins[1]/sum(sum(margins[1]))

Q_1

array([[ 0.71615172,  0.9447375 ,  0.68475257,  0.6543582 ]])

In [79]:
o_11 = bf_table[[0],[0]][0]

e_11 = exp_tab[[0],[0]][0]

r_11 = o_11 - e_11

(o_11, e_11, r_11)

(362, 342.60487314745041, 19.395126852549595)

In [89]:
# Then the standard deviation

res_sd = (exp_tab[[0],[0]][0]* Q_0[[0],[0]][0]*Q_1[[0],[0]][0])**0.5

res_sd

13.075443470258701

In [91]:
Q_1[[0],[0]]

array([ 0.71615172])

In [90]:
# Last, do the division and see if we get the same residual as above

r_11/ res_sd


1.4833245921383542

Let's do it again, more quickly for the two highest residuals, the ones in positions (2, 3) and (3, 2).  

In [93]:
o_23 = bf_table[[2],[3]][0]

e_23 = exp_tab[[2],[3]][0]

r_23 = o_23 - e_23

sd_23 = (exp_tab[[2],[3]][0]* Q_0[[2],[0]][0]*Q_1[[0],[3]][0])**0.5

r_23 / sd_23

4.9623659007898935

In [96]:
o_32 = bf_table[[3],[2]][0]

e_32 = exp_tab[[3],[2]][0]

r_32 = o_32 - e_32

sd_32 = (exp_tab[[3],[2]][0]* Q_0[[3],[0]][0]*Q_1[[0],[2]][0])**0.5

r_32 / sd_32

5.070380065864633